File: Easy/206. Reverse Linked List/RecursiveSolution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        return reverse(head, null);
    }

    private ListNode reverse(ListNode head, ListNode tail) {
        if (head == null)
            return tail;

        ListNode temp = reverse(head.next, head);
        head.next = tail;
        return temp;
    }
}

File: Easy/206. Reverse Linked List/recursive-solution.js
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  return reverse(head, null);

  function reverse(node, tail) {
    if (node == null) return tail;
    let rev = reverse(node.next, node);
    node.next = tail;
    return rev;
  }
};
